rust/src/tftp/tftp.rs

```diff
@@ -0,0 +1,145 @@
+/* Copyright (C) 2017 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+// written by Clément Galland <clement.galland@epita.fr>
+
+extern crate libc;
+extern crate nom;
+
+use std::str;
+use std;
+use std::mem::transmute;
+
+use applayer::LoggerFlags;
+
+#[derive(Debug)]
+pub struct TFTPTransaction {
+    pub opcode : u8,
+    pub filename : String,
+    pub mode : String,
+    pub logged : LoggerFlags
+}
+
+pub struct TFTPState {
+    pub transactions : Vec<TFTPTransaction>
+}
+
+impl TFTPTransaction {
+    pub fn new(opcode : u8, filename : String, mode : String) -> TFTPTransaction {
+        TFTPTransaction {
+            opcode : opcode,
+            filename : filename,
+            mode : mode.to_lowercase(),
+            logged : LoggerFlags::new(),
+        }
+    }
+    pub fn is_mode_ok(&self) -> bool {
+        match self.mode.as_str() {
+            "netascii" | "mail" | "octet" => true,
+            _ => false
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rs_tftp_state_alloc() -> *mut libc::c_void {
+    let state = TFTPState { transactions : Vec::new() };
+    let boxed = Box::new(state);
+    return unsafe{transmute(boxed)};
+}
+
+#[no_mangle]
+pub extern "C" fn rs_tftp_state_free(state: *mut libc::c_void) {
+    let _state : Box<TFTPState> = unsafe{transmute(state)};
+}
+
+#[no_mangle]
+pub extern "C" fn rs_tftp_state_tx_free(state: &mut TFTPState,
+                                        tx_id: libc::uint32_t) {
+    state.transactions.remove(tx_id as usize);
+}
+
+#[no_mangle]
+pub extern "C" fn rs_tftp_get_tx(state: &mut TFTPState,
+                                    tx_id: libc::uint64_t) -> *mut libc::c_void {
+    if state.transactions.len() <= tx_id as usize {
+        return std::ptr::null_mut();
+    }
+    return unsafe{transmute(&state.transactions[tx_id as usize])};
+}
+
+#[no_mangle]
+pub extern "C" fn rs_tftp_get_tx_logged(_state: &mut TFTPState,
+                                        tx: &mut TFTPTransaction,
+                                        logger: libc::uint32_t) -> i8 {
+    if tx.logged.is_logged(logger) {
```

Comment:
this is now outdated, see https://github.com/OISF/suricata/commit/ec62eedc87b13f3f5dbcd845c46b601daed3ca93. Fix is simple, see https://github.com/OISF/suricata/commit/5c6868b327dda74b4f0bb7934923a4a93a2910c7

Reply:
OK understood.